rfcs/45-stageless.md

```diff
@@ -0,0 +1,586 @@
+
+In addition, atomic ordering constraints introduce a new form of schedule unsatisfiability.
+If a system must run between `A` and `B`, but can mutate data that `A` has access to, the schedule is unsatisfiable.
+
+### Flushed ordering constraints
+
+Being able to assert that commands are processed (or state is flushed) between two systems is a critical requirement for being able to specify logically-meaningful ordering of systems without a global view of the schedule.
+
+However, the exact logic involved in doing this is quite complex.
+We need to:
+
+1. **When validating the schedule:** ensure that a flushing system of the appropriate type *could* occur between the pair of systems.
+2. **When executing the schedule:** ensure that a flushing system of the appropriate type *does* occur between the pair of systems.
+
+For this RFC, let's begin with the very simplest strategy: for each pair of systems with a flushed ordering constraint, ensure that a flushing system of the appropriate type is strictly after the first system, and strictly before the second.
+
+We can check if the flushing system is of the appropriate type by checking the type id of the function pointer stored in the `RawSystem`.
+This is cached in the `system_function_type_map` field of the `Schedule` for efficient lookup, yielding the relevant `SystemIds`.
+
+With the ids of the relevant systems in hand, we can check each constraint: for each pair of systems, we must check that at least one of the relevant flushing systems is between them.
+To do so, we need to check both direct *and* transitive strict ordering dependencies, using a depth-first search to walk backwards from the later system using the `CachedOrdering` until a flushing system is reached, and then walking back further from that flushing system until the originating system is reached.
+
+## Drawbacks
+
+1. This will be a massively breaking change to every single Bevy user.
+   1. Any non-trivial control over system ordering will need to be completely reworked.
+   2. Users will typically need to think a bit harder about exactly when they want their gameplay systems to run. In most cases, they should just add the `CoreLabel::AppLogic` label to them.
+2. It will be harder to immediately understand the global structure of Bevy apps.
+   1. Powerful system debugging and visualization tools become even more important.
+3. State transitions are no longer queued up in a stack.
+   1. By default, arbitrarily long chains of state transitions are no longer be processed in the same schedule pass.
+   2. However, users can add as many copies of the `flush_state` system as they would like, and loop it within an exclusive system.
+   3. This also removes "in-stack" and related system groups / logic.
+   4. This can be easily added later, or third-party plugins can create their own abstraction.

```

Comment:
This kind of limiting :(
For single player games it's a nice way to implement pause.
Also in my game I have the following states:

```rust
enum AppState {
    MainMenu,
    Settings,
    DirectConnectMenu,
    CreateGameMenu,
    LobbyMenu,
    InGame,
    InGameMenu,
}
```

Player could go from `MainMenu` to `DirectConnectMenu` and then to `LobbyMenu`. Currently I use `push` to be able to call `pop` in single system that represents my back button.
Also player could go from `InGame` to `InGameMenu` and then also open `Settings` which also plays nicely with `pop`. 
So I would suggest to leave this functionality.

Reply:
On seconds thought, to implement a pause, a resource can be used that says whether the game is paused or not (e.g. combined with run criteria that checks for such resource).
And my use case could be implemented by a separate resource that holds states history.
Maybe it will be even more clear.

Reply:
> And my use case could be implemented by a separate resource that holds states history.
Yep, precisely. We can revisit this if it comes up later: we need to try and reduce the complexity of this RFC wherever we possibly can 😅

Reply:
Yes, you are currently (in existing bevy) forced to try to encode all your system runtime control logic into states, and use the stack, because combining run criteria is hard and unsupported. You can't just add another simple run criteria to check if you are paused or not, or if you are connected to the network, or whatever. You have to find a way to encode everything you need into a state machine.

Stageless will change all that. With this RFC, you can trivially control systems using a combination of multiple different state enums (say, one for menu vs in-game, one for network status, one for pausing, ... whatever) as well as any other run criteria like fixed timestep.

These are the problems this RFC is trying to solve. The new design seems a lot nicer, more flexible, and user friendly, than the state stack. It is doubtful that a state stack will be needed, or desirable at all.